fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStore.java

```diff
@@ -723,6 +725,271 @@ public void updateRecords(@Nonnull Descriptors.FileDescriptor recordsDescriptor)
+        if (messageType.getFieldCount() == 0) {
+            return 1;
+        }
+        return messageType
+                       .getFieldList()
+                       .stream()
+                       .max(Comparator.comparing(DescriptorProtos.FieldDescriptorProto::getNumber))
+                       .get()
+                       .getNumber() + 1;
+    }
+
+    /**
+     * Deprecate a record type from the meta-data.
+     *
+     * @param recordType the record type to be deprecated
+     */
+    public void deprecateRecordType(@Nonnull String recordType) {
+        context.asyncToSync(FDBStoreTimer.Waits.WAIT_DEPRECATE_RECORD_TYPE, deprecateRecordTypeAsync(recordType));
+    }
+
+    /**
+     * Deprecate a record type from the meta-data asynchronously.
+     *
+     * @param recordType the record type to be deprecated
+     * @return a future that completes when the record type is deprecated
+     */
+    @Nonnull
+    public CompletableFuture<Void> deprecateRecordTypeAsync(@Nonnull String recordType) {
+        return loadCurrentProto().thenCompose(metaDataProto -> {
+            RecordMetaDataProto.MetaData.Builder builder = metaDataProto.toBuilder();
+            DescriptorProtos.DescriptorProto.Builder unionBuilder = fetchUnionBuilder(builder.getRecordsBuilder());
+            if (unionBuilder.getName().equals(recordType)) {
+                throw new MetaDataException("Cannot deprecate the union");
+            }
+
+            // deprecate the field from the union.
+            unionBuilder.getFieldBuilderList()
+                    .stream()
+                    .filter(m -> m.getTypeName().equals(recordType))
+                    .findFirst()
+                    .get()
+                    .setOptions(DescriptorProtos.FieldOptions.newBuilder().setDeprecated(true).build());
+
+            RecordMetaDataBuilder recordMetaDataBuilder = createMetaDataBuilder(builder.build());
+            recordMetaDataBuilder.setVersion(metaDataProto.getVersion() + 1);
+            return saveAndSetCurrent(recordMetaDataBuilder.getRecordMetaData().toProto());
+        });
+    }
+
+    /**
+     * Add a field to a record type.
+     *
+     * @param recordType the record type to add the field to
+     * @param field the field to be added
+     */
+    public void addField(@Nonnull String recordType, @Nonnull DescriptorProtos.FieldDescriptorProto field) {
+        context.asyncToSync(FDBStoreTimer.Waits.WAIT_ADD_FIELD, addFieldAsync(recordType, field));
+    }
+
+    /**
+     * Add a field to a record type asynchronously.
+     *
+     * @param recordType the record type to add the field to
+     * @param field the field to be added
+     * @return a future that completes when the record type is deprecated
+     */
+    @Nonnull
+    public CompletableFuture<Void> addFieldAsync(@Nonnull String recordType, @Nonnull DescriptorProtos.FieldDescriptorProto field) {
```

Comment:
> If the field is some message type, does this need to validate that the message exists?

This will fail at `buildFileDescriptor`, right?

> There's another (somewhat more pernicious) possibility that's worse than the field not existing which is the complex type might exist, but it might look different than the field type the user is thinking of. I'm not entirely sure what the right way to fix that is, though.

Yeah, this is hard. But user is anyway asking for a field with a type and I think we should just add the field then.